File: scala-orders/src/main/scala/com/mvv/bank/orders/domain/DomainPrimitives.scala
//noinspection ScalaUnusedSymbol // T O D O: remove after adding test and so on
package com.mvv.bank.orders.domain

import scala.language.strictEquality
//
import scala.annotation.meta.{getter, field, param}
import scala.annotation.unused
import scala.util.matching.Regex
//
import javax.annotation.{Tainted, Untainted}
import javax.annotation.concurrent.Immutable
//
import com.mvv.nullables.{isNotNull, isNull, NullableCanEqualGivens}
import com.mvv.utils.{require, requireNotBlank, requireNotNull, equalImpl}
import com.mvv.collections.in
import com.mvv.log.safe


@Untainted @Immutable
case class Email private (
  @(Tainted @param) @(Untainted @field @getter)
  value: String) derives CanEqual :
  validateEmail(value)
  @Untainted override def toString: String = this.value


object Email extends NullableCanEqualGivens[Email] :
  def apply(@Tainted email: String): Email = new Email(email)

  // standard java methods to get from string. It can help to integrate with other java frameworks.
  def of(@Tainted email: String): Email = Email(email)
  def valueOf(@Tainted email: String): Email = Email(email)



@Untainted @Immutable
case class Phone private (
  @(Tainted @param) @(Untainted @field @getter)
  value: String) derives CanEqual :
  validatePhone(value)
  @Untainted override def toString: String = this.value


object Phone extends NullableCanEqualGivens[Phone] :
  def apply(@Tainted phone: String): Phone = new Phone(phone)

  // standard java methods to get from string. It can help to integrate with other java frameworks.
  def of(@Tainted phone: String): Phone = Phone(phone)
  def valueOf(@Tainted phone: String): Phone = Phone(phone)



// see https://www.baeldung.com/java-email-validation-regex
// RFC5322
// regex "^[a-zA-Z0-9_!#$%&'*+/=?`{|}~^.-]+@[a-zA-Z0-9.-]+$"
// regex "^[a-zA-Z0-9_!#$%&'*+/=?`{|}~^-]+(?:\\.[a-zA-Z0-9_!#$%&'*+/=?`{|}~^-]+)*@"
//        + "[a-zA-Z0-9-]+(?:\\.[a-zA-Z0-9-]+)*$")
private val emailOwaspPattern = Regex("^[a-zA-Z0-9_+&*-]+(?:\\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,7}$")
private def validateEmail(@Tainted email: String): Unit =
  requireNotBlank(email, "Email cannot be null/blank.")
  if (!emailOwaspPattern.matches(email.nn)) throw IllegalArgumentException(s"Invalid email [${email.safe}].")


// see https://www.baeldung.com/java-regex-validate-phone-numbers
// regex "^(\\+\\d{1,3}( )?)?((\\(\\d{3}\\))|\\d{3})[- .]?\\d{3}[- .]?\\d{4}$"
// regex "^(\\+\\d{1,3}( )?)?((\\(\\d{1,3}\\))|\\d{1,3})[- .]?\\d{3,4}[- .]?\\d{4}$"
private val phonePattern = Regex("^\\+?[1-9][0-9]{7,14}$")
private def validatePhone(@Tainted phone: String): Unit =
  requireNotBlank(phone, "Phone number cannot be null/blank.")
  if !phonePattern.matches(phone.nn) then throw IllegalArgumentException(s"Invalid phone number [${phone.safe}].")

File: kotlin-orders/kotlin-orders-dto/src/main/java/com/mvv/bank/orders/domain/DomainPrimitives.kt
package com.mvv.bank.orders.domain

import com.mvv.bank.log.safe
import javax.annotation.Tainted
import javax.annotation.Untainted
import javax.annotation.concurrent.Immutable


@Untainted @Immutable
class Email private constructor (@param:Tainted @field:Untainted val value: String) {
    init { validateEmail(value) }
    @Untainted
    override fun toString(): String = value
    override fun equals(other: Any?): Boolean =
        (this === other) || ((this.javaClass == other?.javaClass) && (this.value == (other as Email).value))
    override fun hashCode(): Int = value.hashCode()

    companion object {
        operator fun invoke(email: String) = Email(email)

        // for Java (MapStruct so on)
        @JvmStatic fun of(email: String) = invoke(email)
        // standard java method to get from string. It can help to integrate with other java frameworks.
        @JvmStatic fun valueOf(email: String) = invoke(email)
    }
}


@Untainted @Immutable
class Phone private constructor (@param:Tainted @field:Untainted val value: String) {
    init { validatePhone(value) }
    @Untainted
    override fun toString(): String = value
    override fun equals(other: Any?): Boolean =
        (this === other) || ((this.javaClass == other?.javaClass) && (this.value == (other as Phone).value))
    override fun hashCode(): Int = value.hashCode()

    companion object {
        operator fun invoke(email: String) = Phone(email)

        // for java (MapStruct, so on)
        @JvmStatic fun of(email: String) = invoke(email)
        // standard java method to get from string. It can help to integrate with other java frameworks.
        @JvmStatic fun valueOf(email: String) = invoke(email)
    }
}


// see https://www.baeldung.com/java-email-validation-regex
// RFC5322
//private val emailRfc5322Pattern = Regex("^[a-zA-Z0-9_!#$%&'*+/=?`{|}~^.-]+@[a-zA-Z0-9.-]+$")
// Strict Regular Expression Validation
//private val emailStrictPattern = Regex("^[a-zA-Z0-9_!#$%&'*+/=?`{|}~^-]+(?:\\.[a-zA-Z0-9_!#$%&'*+/=?`{|}~^-]+)*@"
//        + "[a-zA-Z0-9-]+(?:\\.[a-zA-Z0-9-]+)*$")
private val emailOwaspPattern = Regex("^[a-zA-Z0-9_+&*-]+(?:\\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,7}$")
private fun validateEmail(email: CharSequence?) {
    check(!email.isNullOrBlank()) { "Email cannot be null/blank." }
    if (!emailOwaspPattern.matches(email)) throw IllegalArgumentException("Invalid email [${email.safe}].")
}


// see https://www.baeldung.com/java-regex-validate-phone-numbers
//private val phonePattern = Regex("^(\\+\\d{1,3}( )?)?((\\(\\d{3}\\))|\\d{3})[- .]?\\d{3}[- .]?\\d{4}$")
//private val phonePattern = Regex("^(\\+\\d{1,3}( )?)?((\\(\\d{1,3}\\))|\\d{1,3})[- .]?\\d{3,4}[- .]?\\d{4}$")
private val phonePattern = Regex("^\\+?[1-9][0-9]{7,14}$")
private fun validatePhone(email: CharSequence?) {
    check(!email.isNullOrBlank()) { "Phone number cannot be null/blank." }
    if (!phonePattern.matches(email)) throw IllegalArgumentException("Invalid phone number [${email.safe}].")
}


/*
private const val MAX_ACCOUNT_LENGTH = 50
private val ACCOUNT_ID_PATTERN = Regex("^[a-zA-Z0-9]*$")

private fun validateAccountId(accountId: String) {
    val isValid = accountId.isNotBlank()
            && accountId.length <= MAX_ACCOUNT_LENGTH
            && accountId.matches(ACCOUNT_ID_PATTERN)

    if (!isValid) {
        throw InvalidArgumentException(
            "Account ID ${accountId.safe} has wrong format.",
            accountId,
            "Account ID should be string of chars/digits with 1-${MAX_ACCOUNT_LENGTH} length.")
    }
}
*/
